# Exploring Hacker News Posts

In this project, we'll work with a data set of submissions to a very popular technology site, Hacker News. This is a website where posts made by users are voted and get commented upon, very similar in functionality to Reddit. Posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result. 

We'll be working with two types of posts whose titles begin with: Ask HN and Show HN. Ask HN posts are concerned with asking the community a specific question while Show HN posts are concerned with showing the community a project or a product. From comparison of these two type of posts, we'll determine the following:
- Which type of post gets more comments on average?
- Do posts created at a certain time receive more comments on average?

In [1]:
from csv import reader
opened_file=open('hacker_news.csv')
read_file=reader(opened_file)
hn=list(read_file)
print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers=hn[0]
hn=hn[1:]
print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


# Extracting Ask HN and Show HN Posts

Here, we are separating the posts beginning with `Ask HN` and `Show HN` (taking in consideration the case variations) into two different lists. We're also sorting all the other type of posts in the other_posts list and finally, checking the number of posts in each list.

In [3]:
ask_posts=[]
show_posts=[]
other_posts=[]

for row in hn:
    title=row[1]
    lower_title=title.lower()
    if lower_title.startswith('ask hn'):
        ask_posts.append(row)
    elif lower_title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
print("Number of ask posts is " + str(len(ask_posts)))
print("Number of show posts is " + str(len(show_posts)))
print("Number of other posts is " + str(len(other_posts)))

Number of ask posts is 1744
Number of show posts is 1162
Number of other posts is 17194


# Average Number of Comments

Here, we see which one of the two: ask posts or show posts receive more comments on average.

In [4]:
total_ask_comments=0
for post in ask_posts:
    comfreq=int(post[4])
    total_ask_comments+=comfreq
avg_ask_comments=total_ask_comments/len(ask_posts)
print("Average number of comments on ask posts is " + str(avg_ask_comments))

total_show_comments=0
for post in show_posts:
    comfreq=int(post[4])
    total_show_comments+=comfreq
avg_show_comments=total_show_comments/len(show_posts)
print("Average number of comments on show posts is " + str(avg_show_comments))

Average number of comments on ask posts is 14.038417431192661
Average number of comments on show posts is 10.31669535283993


From the above calculation, we can say that ask posts get a higher number of comments than show posts. They get about 4 more comments per post than that of show posts.

# Finding the Amount of Ask Posts and Comments by Hour Created

In [5]:
import datetime as dt
result_list=[]
for post in ask_posts:
    time=post[-1]
    com=int(post[4])
    result_list.append([time, com])

counts_by_hour={}
comments_by_hour={}
date_format = "%m/%d/%Y %H:%M"

for post in result_list:
    date_and_time=post[0]
    comment=post[1]
    hour=dt.datetime.strptime(date_and_time, date_format).strftime("%H")
    
    if hour not in counts_by_hour:
        counts_by_hour[hour]=1
        comments_by_hour[hour]=comment
    else:
        counts_by_hour[hour]+=1
        comments_by_hour[hour]+=comment

# Calculating the Average Number of Comments for Ask HN Posts by Hour

We will now create a list of lists containing the hours during which posts were created and the average number of comments those posts received. Finally, we'll display this list in the end.

In [6]:
avg_by_hour=[]

for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr]/counts_by_hour[hr]])
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]

# Sorting and Printing Values from a List of Lists

In [7]:
swap_avg_by_hour=[]
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
swap_avg_by_hour
sorted_swap=sorted(swap_avg_by_hour, reverse=True)
print("Top 5 Hours for Ask Posts Comments")
for avg, hr in sorted_swap[:5]:
    print("{}: {:.2f} average comments per post".format(
            dt.datetime.strptime(hr, "%H").strftime("%H:%M"),
            avg))

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


From our above analysis, we see that posts created at 15:00 get the highest number of comments by a large margin, with an average of 38.59 comments per post. Hours like 02:00 and 20:00 can also be favourable to post on Hacker News.

# Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00.